algorithms.py

```diff
@@ -427,20 +520,228 @@ def hudson_simulate(self):
+                    self.common_ancestor_event(ca_population,0)
                 else:
                     # print("MIG EVENT")
                     self.migration_event(mig_source, mig_dest)
         return self.finalise()
+    
+    def hudson_simulate_until(self,max_time):
+        """
+        Simulates the algorithm until all loci have coalesced or
+        max_time is reached. returns time on exit.
+        tables NOT finalized
+        """
+        infinity = sys.float_info.max
+        #only worried about label 0 below
+        while sum(pop.get_num_ancestors() for pop in self.P[0]) != 0 and self.t < max_time:
+            self.verify()
+            rate = self.r * self.L[0].get_total()
+            t_re = infinity
+            if rate != 0:
+                t_re = random.expovariate(rate)
+            # Common ancestor events occur within demes.
+            t_ca = infinity
+            for index, pop in enumerate(self.P[0]):
+                t = pop.get_common_ancestor_waiting_time(self.t)
+                if t < t_ca:
+                    t_ca = t
+                    ca_population = index
+            t_mig = infinity
+            # Migration events happen at the rates in the matrix.
+            for j in range(len(self.P[0])):
+                source_size = self.P[0][j].get_num_ancestors()
+                for k in range(len(self.P[0])):
+                    rate = source_size * self.migration_matrix[j][k]
+                    if rate > 0:
+                        t = random.expovariate(rate)
+                        if t < t_mig:
+                            t_mig = t
+                            mig_source = j
+                            mig_dest = k
+            min_time = min(t_re, t_ca, t_mig)
+            assert min_time != infinity
+            if self.t + min_time > max_time:
+                self.t = max_time
+                return(max_time)
+            elif self.t + min_time > self.modifier_events[0][0]:
+                t, func, args = self.modifier_events.pop(0)
+                self.t = t
+                func(*args)
+            else:
+                self.t += min_time
+                if min_time == t_re:
+                    # print("RE EVENT")
+                    self.hudson_recombination_event(0)
+                elif min_time == t_ca:
+                    # print("CA EVENT")
+                    self.common_ancestor_event(ca_population,0)
+                else:
+                    # print("MIG EVENT")
+                    self.migration_event(mig_source, mig_dest)
+        return self.t
+   
+    def single_sweep_simulate(self):
+        """
+        Simulates a single sweep model using
+        phases of coalescent and structured coalsct
+        """
+        
+        assert( True in (["single_sweep_event" in x for x in self.modifier_events]))
+        #pull out sweep
+        idx = ["single_sweep_event" in x for x in self.modifier_events].index(True)
```

Comment:
I think we'll probably want to return to this and add the ``ModelChangeEvent`` to this inferface later, but this is good for now.